drivers/gpio/gpio_stmpe1600.c

```diff
@@ -0,0 +1,288 @@
+	}
+	return ret;
+}
+
+static int read_reg16(const struct stmpe1600_config * const config, uint8_t reg, uint16_t *value)
+{
+	int ret;
+	uint16_t transfer_data;
+
+	LOG_DBG("STMPE1600[0x%02X]: read REG[0x%02X..0x%02X]",
+		  config->i2c_slave_addr, reg, reg + 1);
+
+	ret = i2c_burst_read(config->i2c_bus, config->i2c_slave_addr, reg,
+			     (uint8_t *) &transfer_data, sizeof(transfer_data));
+
+	if (ret != 0) {
+		LOG_ERR("STMPE1600[0x%02X]: read error REG[0x%02X..0x%02X]: %d",
+			config->i2c_slave_addr, reg, reg + 1, ret);
+	} else {
+		*value = sys_le16_to_cpu(transfer_data);
+		LOG_DBG("STMPE1600[0x%02X]: read REG[0x%02X..0x%02X] => %04x",
+			config->i2c_slave_addr, reg, reg + 1, *value);
+	}
+	return ret;
+}
+
+static int set_pin_dir(const struct device *dev, gpio_pin_t pin, gpio_flags_t flags)
+{
+	int ret;
+	bool set_state = false;
+	struct stmpe1600_drvdata *const drvdata = (struct stmpe1600_drvdata *const) dev->data;
+
+	if ((flags & GPIO_INPUT) != 0U) {
+		drvdata->GPDR &= ~BIT(pin);
+	} else {
+		drvdata->GPDR |= BIT(pin);
+		if ((flags & GPIO_OUTPUT_INIT_HIGH) != 0U) {
+			drvdata->GPSR |= BIT(pin);
+			set_state = true;
+		} else if ((flags & GPIO_OUTPUT_INIT_LOW) != 0U) {
+			drvdata->GPSR &= ~BIT(pin);
+			set_state = true;
+		}
+	}
+
+	ret = write_reg16(dev->config, REG_GPDR_LSB, drvdata->GPDR);
+	if (ret != 0) {
+		return ret;
+	}
+	if (set_state) {
+		ret = write_reg16(dev->config, REG_GPSR_LSB, drvdata->GPSR);
+	}
+	return ret;
+}
+
+static int stmpe1600_configure(const struct device *dev,
+			       gpio_pin_t pin, gpio_flags_t flags)
+{
+	int ret;
+	struct stmpe1600_drvdata *const drvdata = (struct stmpe1600_drvdata *const) dev->data;
+	const struct stmpe1600_config *const config = dev->config;
+
+	/* No support for disconnected pin */
+	if ((flags & (GPIO_INPUT | GPIO_OUTPUT)) == GPIO_DISCONNECTED) {
+		return -ENOTSUP;
+	}
+
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	k_sem_take(&drvdata->lock, K_FOREVER);
+
+	ret = set_pin_dir(dev, pin, flags);
+	if (ret != 0) {
+		LOG_ERR("STMPE1600[0x%X]: error setting pin direction (%d)",
+			config->i2c_slave_addr, ret);
+	}
+
+	k_sem_give(&drvdata->lock);
+	return ret;
+}
+
+static int stmpe1600_port_get_raw(const struct device *dev, uint32_t *value)
+{
+	int ret;
+	uint16_t reg_value;
+	struct stmpe1600_drvdata *const drvdata = (struct stmpe1600_drvdata *const) dev->data;
+
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	k_sem_take(&drvdata->lock, K_FOREVER);
+	ret = read_reg16(dev->config, REG_GPMR_LSB, &reg_value);
+	k_sem_give(&drvdata->lock);
+
+	if (ret == 0) {
+		*value = reg_value;
+	}
+
+	return ret;
+}
+
+static int stmpe1600_port_set_masked_raw(const struct device *dev,
+					 uint32_t mask, uint32_t value)
+{
+	int ret;
+	struct stmpe1600_drvdata *const drvdata = (struct stmpe1600_drvdata *const)dev->data;
+
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	drvdata->GPSR = (drvdata->GPSR & ~mask) | (mask & value);
```

Comment:
We can update value of `drvdata->GPSR` only if we successfully written it to the device, i.e. only if call to `write_reg16` returned 0. You also need to be careful about the usage of semaphores.